modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v4_0_0/UpgradeKaleoDefinitionVersion.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.workflow.kaleo.internal.upgrade.v4_0_0;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.dao.orm.common.SQLTransformer;
+import com.liferay.portal.kernel.upgrade.UpgradeException;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.workflow.kaleo.internal.upgrade.v4_0_0.util.KaleoDefinitionVersionTable;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * @author Jesse Yeh
+ */
+public class UpgradeKaleoDefinitionVersion extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		StringBundler sb = new StringBundler(3);
+
+		sb.append("select kaleoDefinitionVersionId, name, version, ");
+		sb.append("kaleoDefinitionId from KaleoDefinitionVersion order by ");
+		sb.append("name, LENGTH(version), cast(version as decimal(4, 2)) asc");
```

Comment:
Do you know if "decimal(4,2)" is a valid syntax for all our supported databases? We don't have a transformation for that in BaseSQLTransformerLogic.java or DB2/Oracle/MySQL/PostgreSQL/SQLServerSQLTransformerLogic so it must be valid syntax for all databases.

Reply:
It seems `cast(version as decimal(4, 2))` conforms to the SQL standard, so it should be portable across different databases. To test this, I used the following query and [this validator](https://developer.mimer.com/sql-2016-validator/):
```
select * from foo order by cast(bar as decimal(4,2))
```

Alternatively, I'm wondering if it might be better to use `abs(version)`, which seems to be supported by the different databases in our [7.3 compatibility matrix](https://help.liferay.com/hc/en-us/articles/360049238151-Liferay-DXP-7-3-Liferay-Commerce-3-0-Compatibility-Matrix) and fulfills the same purpose for sorting.

Reply:
Based on the following documentation, it looks like it should be valid syntax:

DB2: https://www.ibm.com/docs/en/db2-for-zos/11?topic=expressions-cast-specification
Maria DB: https://mariadb.com/kb/en/cast/
MySQL: https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#cast-function-descriptions (see `DECIMAL` entry)
Oracle: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Data-Types.html#GUID-0BC16006-32F1-42B1-B45E-F27A494963FF (`DECIMAL` is equivalent to the `NUMBER` type)
Postgresql: https://www.postgresql.org/docs/12/datatype.html (`decimal` is an alias for `numeric`)
SQL Server: https://docs.microsoft.com/en-us/sql/t-sql/data-types/decimal-and-numeric-transact-sql?view=sql-server-ver15

Alternatively, I'm wondering if it might be better to instead use `abs(version)`, which seems to also be supported by the different databases.